### ISSUE TITLE
Istio 1.13 to 1.14 upgrade issue - SPIFFE errors

### ISSUE THREAD

**Issue by kalyanteja19** [istio/istio#42618]
### Bug Description

Hi Team,

We are upgrading from istio 1.13 to 1.14 issue and see that pods are not in ready state, facing errors related to directory for ./var/run/secrets/workload-spiffe-uds/socket, as it does not have permission to create directory.

**Installation method:** We are using istio operator profile to deploy istiod into cluster.

We observed that sidecar injection template is not adding the volumes related to SPIFFE integration.

Pods are stuck as sidecar is not getting ready because of these issues. Kindly assist to resolve this issue and suggest if the SPIFFE integration is mandatory for all istio installations. If optional, how to disable it.

@howardjohn @hzxuzhonghu @GregHanson @ZhengzheYang , kindly assist, as this is priority for us

### Version

```prose
istioctl version                                                                                                                                                                      client version: 1.14.6
control plane version: 1.14.6
data plane version: 1.14.6 (9 proxies), 1.13.9 (9 proxies)

kubectl version --short                                                                                                                                                                Client Version: v1.23.12
Server Version: v1.23.12
```

### Additional Information

Below are the error logs for reference.

Below is the error for reference.

2022-12-30T05:41:16.447287Z     info    JWT policy is third-party-jwt
2022-12-30T05:41:16.447291Z     info    using credential fetcher of JWT type in cluster.local trust domain
2022-12-30T05:41:16.447564Z     info    platform detected is Azure
2022-12-30T05:41:16.450341Z     warn    HTTP request unsuccessful with status: 400 Bad Request
2022-12-30T05:41:16.450638Z     warn    HTTP request unsuccessful with status: 400 Bad Request
2022-12-30T05:41:16.450744Z     info    SDS socket not found. Starting Istio SDS Server
2022-12-30T05:41:16.450761Z     info    CA Endpoint istiod.istio-system.svc:15012, provider Citadel
2022-12-30T05:41:16.450783Z     info    Using CA istiod.istio-system.svc:15012 cert with certs: var/run/secrets/istio/root-cert.pem
2022-12-30T05:41:16.450857Z     info    citadelclient   Citadel client using custom root cert: var/run/secrets/istio/root-cert.pem
2022-12-30T05:41:16.453724Z     info    Opening status port 15020
2022-12-30T05:41:16.476804Z     info    ads     All caches have been synced up in 38.204908ms, marking server ready
2022-12-30T05:41:16.477067Z     warn    Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2022-12-30T05:41:16.477134Z     error   sds     Failed to set up UDS path: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/work
load-spiffe-uds/socket: bind: no such file or directory
2022-12-30T05:41:16.477164Z     info    sds     SDS server for workload certificates started, listening on "./var/run/secrets/workload-spiffe-uds/socket"
2022-12-30T05:41:16.477189Z     info    xdsproxy        Initializing with upstream address "istiod.istio-system.svc:15012" and cluster "Kubernetes"
2022-12-30T05:41:16.477238Z     info    sds     Starting SDS grpc server
2022-12-30T05:41:16.477294Z     warn    Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2022-12-30T05:41:16.477327Z     error   sds     SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": li
sten unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-30T05:41:16.479074Z     info    Pilot SAN: [istiod.istio-system.svc]
2022-12-30T05:41:16.481135Z     info    Starting proxy agent
2022-12-30T05:41:16.481168Z     info    Epoch 0 starting
2022-12-30T05:41:16.481192Z     info    Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --drain-strategy immediate --parent-shutdown-time-s 60 --local
-address-ip-version v4 --file-flush-interval-msec 1000 --disable-hot-restart --log-format %Y-%m-%dT%T.%fZ       %l      envoy %n        %v -l warning --component-log-level misc:error --conc
urrency 2]
2022-12-30T05:41:16.481390Z     info    starting Http service at 127.0.0.1:15004
2022-12-30T05:41:16.700688Z     info    xdsproxy        connected to upstream XDS server: istiod.istio-system.svc:15012
2022-12-30T05:41:16.815161Z     info    cache   generated new workload certificate      latency=337.926565ms ttl=23h59m59.184854911s
2022-12-30T05:41:16.815214Z     info    cache   Root cert has changed, start rotating root cert
2022-12-30T05:41:16.815239Z     info    ads     XDS: Incremental Pushing:0 ConnectedEndpoints:0 Version:
2022-12-30T05:41:16.815305Z     info    cache   returned workload trust anchor from cache       ttl=23h59m59.184710011s
2022-12-30T05:41:17.477643Z     warn    Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2022-12-30T05:41:17.477751Z     error   sds     SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": li
sten unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-30T05:41:19.478961Z     warn    Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2022-12-30T05:41:19.479043Z     error   sds     SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": li
sten unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-30T05:41:23.479278Z     warn    Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2022-12-30T05:41:23.479385Z     error   sds     SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": li
sten unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory

### Affected product area

- [ ] Ambient
- [x] Docs
- [X] Installation
- [x] Networking
- [x] Performance and Scalability
- [ ] Extensions and Telemetry
- [x] Security
- [x] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [X] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### Is this the right place to submit this?

- [X] This is not a security vulnerability
- [X] This is not a question about how to use Istio

**Comment by kalyanteja19** [comment]
Hi Team, 

We observed that below volume mounts and volumes are not getting adding to istio-proxy sidecar, hence they are stuck. Please assist. 

            volumeMounts:
            - name: workload-socket
              mountPath: /var/run/secrets/workload-spiffe-uds
            - name: workload-certs
              mountPath: /var/run/secrets/workload-spiffe-credentials
          volumes:
          - emptyDir:
            name: workload-socket
          - emptyDir:
            name: workload-certs

**Comment by Pawan-Bishnoi** [comment]
These should be there in your sidecar template. Empty directory works.

**Comment by kalyanteja19** [comment]
> These should be there in your sidecar template. Empty directory works.

Thank you so much. It seems these volumes should be added for all sidecars in istio 1.14, though we donot use any external trush domain. Can you please confirm?

**Comment by Pawan-Bishnoi** [comment]
Yes, afaik. I am not an expert though but that's been my observation.
Adding these mounts to the sidecar injector template should unblock you. While you wait for someone with more experience to comment.

**Comment by kalyanteja19** [comment]
> @howardjohn @hzxuzhonghu @GregHanson @ZhengzheYang

Thank you.

@howardjohn @hzxuzhonghu @GregHanson @ZhengzheYang , after we upgrade from istio 1.13 to 1.14.6, sidecars are not getting ready, as they are not deployed with the below. Kindly help.

        volumeMounts:
        - name: workload-socket
          mountPath: /var/run/secrets/workload-spiffe-uds
        - name: workload-certs
          mountPath: /var/run/secrets/workload-spiffe-credentials
      volumes:
      - emptyDir:
        name: workload-socket
      - emptyDir:
        name: workload-certs

**Comment by howardjohn** [comment]
Why is the volume not there? It's part of the injection config. Have you customized it?

**Comment by kalyanteja19** [comment]
> howardjohn

@howardjohn Thanks for your reply. Yeah, these should be injected as per sidecar config. But those volumes are not getting added into sidecar istio-proxy.

**Comment by kalyanteja19** [comment]
@howardjohn , Thank you for looking into this issue.

Issue occurred because of old version envoyfilters. After removing old version envoy filters, configmap updated properly and pods are running with istio proxies.

When we are upgrading istio, by default old envoy filters are not removing but new ones are getting added. Hope there should not be any need to keep old version envoy filters, please confirm.

**Comment by howardjohn** [comment]
Old telemetry envoy filters should not be relevant for newer versions. Looks like this is resolved

**Comment by fedeostrit** [comment]
hello everyone, this error is still present, the infrastructure I have is deployed with official istio charts (Istiod and istioingress) the istio/pilot and istio/proxyv2 image was updated, the CRDs were updated with istiod and the images of the gateways were changed (istio/proxyv2:1.17.1) but when they try to raise the gateways they throw the error:

info sds Starting SDS grpc server
2023-03-07T21:07:49.093844Z warn Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets/workload-spiffe-uds: read-only file system
2023-03-07T21:07:49.093968Z error sds SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory


to install istiod the chart https://github.com/istio/manifests/charts/istio-control/istio-discovery is used and the image is downloaded from dockerhub and pushed to a private registry (in this case istio/pilot: 1.17.1)


to install the gateway the chart https://github.com/istio/manifests/charts/gateways/istio-ingress is used, the image is taken from dockerhub and pushed to a private registry (istio/proxyv2)


version 1.13.3 with this installation works with the upgrade it doesn't work anymore because the gateways don't get up with the errors that have already been added

**Comment by fedeostrit** [comment]
I think the problem is in the templates of the chart https://github.com/istio/istio/tree/master/manifests/charts/gateways/istio-ingress/templates, apparently they are not generating the mountpoint for that volume, it would have I have to do an update for the charts that are in these repos because they apparently were outdated

**Comment by fedeostrit** [comment]
Problem solved, I was using the repo chart of version 1.13.3 instead of the version 1.17.1 chart, sorry for this query
